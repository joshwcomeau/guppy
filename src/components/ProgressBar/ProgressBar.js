// @flow
import React, { Component } from 'react';
import styled from 'styled-components';
import { Spring, animated } from 'react-spring';

import { GRADIENTS } from '../../constants';

// TODO: consider renaming stiffness and damping to tension and friction
type Props = {
  height: number,
  progress: number,
  stiffness: number,
  damping: number,
  colors: Array<string>,
  reset: boolean,
};

class ProgressBar extends Component<Props> {
  static defaultProps = {
    height: 8,
    stiffness: 32,
    damping: 32,
    colors: GRADIENTS.progress,
    reset: false,
  };

  render() {
    const { height, progress, stiffness, damping, colors, reset } = this.props;

    return (
      <Wrapper height={height}>
        <Spring
          from={{ progress: 0 }}
          to={{ progress }}
          config={{ tension: stiffness, friction: damping }}
          native
          reset={reset}
        >
          {interpolated => (
            <ProgressGradient
              colors={colors}
              progress={interpolated.progress}
            />
          )}
        </Spring>
      </Wrapper>
    );
  }
}

export const Wrapper = styled.div`
  position: relative;
  height: ${props => props.height}px;
  width: 100%;
`;

const ProgressGradient = animated(styled.div.attrs({
  style: props => ({
    clipPath: `polygon(
      0% 0%,
      ${props.progress * 100}% 0%,
      ${props.progress * 100}% 100%,
      0% 100%
    `,
  }),
})`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: linear-gradient(to right, ${props => props.colors.join(', ')});
`);

export default ProgressBar;
