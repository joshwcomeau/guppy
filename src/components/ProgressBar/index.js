// @flow
export { default } from './ProgressBar';
