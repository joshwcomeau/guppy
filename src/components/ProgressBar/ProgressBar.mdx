---
name: ProgressBar
menu: Reusable Components
---

import { Playground, Props } from 'docz';
import ProgressBar from './ProgressBar';
import ProgressManager from '../../docz/components/ProgressManager';
import { RAW_COLORS } from '../../constants';

# ProgressBar

<Props of={ProgressBar} />

### Empty
<Playground>
  <ProgressBar progress={0} />
</Playground>

### Half full
<Playground>
  <ProgressBar progress={0.5} />
</Playground>

### Full
<Playground>
  <ProgressBar progress={1} />
</Playground>

### Dynamic
<Playground>
  <ProgressManager>
    {({ progress }) => <ProgressBar progress={progress} />}
  </ProgressManager>
</Playground>

### Dynamic (tighter)
<Playground>
  <ProgressManager>
    {({ progress }) => (
      <ProgressBar progress={progress} stiffness={170} damping={16} />
    )}
  </ProgressManager>
</Playground>

### Dynamic (tall + rainbow)
<Playground>
  <ProgressManager>
    {({ progress }) => (
      <ProgressBar
        height={16}
        progress={progress}
        colors={[
          RAW_COLORS.red[500],
          RAW_COLORS.orange[500],
          RAW_COLORS.lime[500],
          RAW_COLORS.teal[500],
          RAW_COLORS.blue[700],
          RAW_COLORS.purple[500],
        ]}
      />
    )}
  </ProgressManager>
</Playground>
