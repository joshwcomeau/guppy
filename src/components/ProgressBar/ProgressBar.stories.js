// @flow
import React, { Fragment, Component } from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import { RAW_COLORS } from '../../constants';

import Showcase from '../../../.storybook/components/Showcase';
import { StrokeButton } from '../Button';
import ProgressBar from './ProgressBar';

type Props = { children: (data: any) => React$Node };
type State = { progress: number };

class ProgressManager extends Component<Props, State> {
  state = {
    progress: 0.5,
  };

  updateProgress = () => {
    const { progress } = this.state;

    const nextProgress = progress === 1 ? 0 : progress + 0.25;

    this.setState({ progress: nextProgress });
  };

  render() {
    const { children } = this.props;
    const { progress } = this.state;

    return (
      <Fragment>
        {children({
          progress,
          updateProgress: this.updateProgress,
        })}
        <br />
        <StrokeButton size="small" onClick={this.updateProgress}>
          Generate new value
        </StrokeButton>
      </Fragment>
    );
  }
}

storiesOf('ProgressBar', module).add(
  'default',
  withInfo()(() => (
    <Fragment>
      <Showcase label="Empty">
        <ProgressBar progress={0} />
      </Showcase>

      <Showcase label="Half full">
        <ProgressBar progress={0.5} />
      </Showcase>

      <Showcase label="Full">
        <ProgressBar progress={1} />
      </Showcase>

      <Showcase label="Dynamic">
        <ProgressManager>
          {({ progress }) => <ProgressBar progress={progress} />}
        </ProgressManager>
      </Showcase>

      <Showcase label="Dynamic (tighter)">
        <ProgressManager>
          {({ progress }) => (
            <ProgressBar progress={progress} stiffness={170} damping={16} />
          )}
        </ProgressManager>
      </Showcase>

      <Showcase label="Dynamic (tall + rainbow)">
        <ProgressManager>
          {({ progress }) => (
            <ProgressBar
              height={16}
              progress={progress}
              colors={[
                RAW_COLORS.red[500],
                RAW_COLORS.orange[500],
                RAW_COLORS.lime[500],
                RAW_COLORS.teal[500],
                RAW_COLORS.blue[700],
                RAW_COLORS.purple[500],
              ]}
            />
          )}
        </ProgressManager>
      </Showcase>
    </Fragment>
  ))
);
