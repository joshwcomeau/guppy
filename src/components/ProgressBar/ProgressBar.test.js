/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { shallow, mount } from 'enzyme';
import ProgressBar, { Wrapper } from './ProgressBar';

describe('ProgressBar component', () => {
  describe('should render', () => {
    // We could do the tests in a loop but for three test cases
    // it's OK to duplicate code
    it('with 0%', () => {
      const wrapper = shallow(<ProgressBar progress={0} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('with 50%', () => {
      const wrapper = shallow(<ProgressBar progress={0.5} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('with 100%', () => {
      const wrapper = shallow(<ProgressBar progress={1.0} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should apply height style to Wrapper', () => {
    const wrapper = mount(<ProgressBar height={10} progress={0.5} />);

    expect(wrapper.find(Wrapper)).toHaveStyleRule('height', '10px');
  });
});
