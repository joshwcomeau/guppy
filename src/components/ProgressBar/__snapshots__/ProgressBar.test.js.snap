// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar component should render with 0% 1`] = `
<styled.div
  height={8}
>
  <Spring
    config={
      Object {
        "friction": 32,
        "tension": 32,
      }
    }
    from={
      Object {
        "progress": 0,
      }
    }
    hold={false}
    immediate={false}
    impl={[Function]}
    inject={[Function]}
    native={true}
    reset={false}
    to={
      Object {
        "progress": 0,
      }
    }
  >
    <Component />
  </Spring>
</styled.div>
`;

exports[`ProgressBar component should render with 50% 1`] = `
<styled.div
  height={8}
>
  <Spring
    config={
      Object {
        "friction": 32,
        "tension": 32,
      }
    }
    from={
      Object {
        "progress": 0,
      }
    }
    hold={false}
    immediate={false}
    impl={[Function]}
    inject={[Function]}
    native={true}
    reset={false}
    to={
      Object {
        "progress": 0.5,
      }
    }
  >
    <Component />
  </Spring>
</styled.div>
`;

exports[`ProgressBar component should render with 100% 1`] = `
<styled.div
  height={8}
>
  <Spring
    config={
      Object {
        "friction": 32,
        "tension": 32,
      }
    }
    from={
      Object {
        "progress": 0,
      }
    }
    hold={false}
    immediate={false}
    impl={[Function]}
    inject={[Function]}
    native={true}
    reset={false}
    to={
      Object {
        "progress": 1,
      }
    }
  >
    <Component />
  </Spring>
</styled.div>
`;
